broadinstitute/single_cell_portal_core
# Restore "Create annotations" (SCP-2956)

This restores, refines, and refactors the "[Create annotations](https://github.com/broadinstitute/single_cell_portal/wiki/Annotations)" feature.  

Due to a refactor done for maintainable spatial transcriptomics support, the "Create annotations" front-end broke.  It had used a JS-in-ERB structure and fetched dynamic `script` tags.  That approach makes front-end debugging and development hard.  Rather than fixing those bugs by adding substantial new code to cement long-term support for the legacy way, I decided to fix the issues by extending the refactor.

With these changes, "Create annotations" is not only fixed but also more maintainable.  Syntax highlighting, definition lookup, and JavaScript breakpoints are trivial.  Creating user annotations is done via JSON REST API.  The JavaScript -- which is more inherently complex than usual -- now has clearer and more consistent names, smaller and more testable functions, and more high-level code explanation for maintainers.

Demo:

https://user-images.githubusercontent.com/1334561/105666592-99558480-5ea7-11eb-8bc3-f61a3ebf61c0.mov

To test:
1. Sign in to SCP
2. Go to "Explore" tab in Study Overview for a study that has visualizations fully initialized, e.g. [SCP197 on staging](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP197) or [SCP1 on production](https://singlecell.broadinstitute.org/single_cell/study/SCP1).
3. Click hamburger menu to left of search box
4. Click "Create Annotations", then "Select Cells"
5. In the "Clusters" scatter plot, drag cursor to trace an outline of cells of interest
6. Enter text into inputs for "Name this annotation" box, and "Label these cells" boxes for both selections (i.e. Selected and Unselected)
7. Click "Create Annotations"
8. Confirm that success message appears, after waiting for annotation to save
9. Click "View Options" at upper right
10. Click "Select annotation" dropdown menu
11. Confirm the text entered for "Name this annotation" appears as an option under the "User annotation" option group
12. Pick your user annotation from the menu
13. Confirm your annotation appears in the scatter plot, along with your two annotation labels

This satisfies SCP-2956.
